### ISSUE TITLE
--source not behaving as I expected

### ISSUE THREAD

**Issue by Ark-kun** [newren/git-filter-repo#66]
I need to extract couple of directories from a monorepo and also make one replacement.
Here is the script I've used initially:

```
git clone --bare monorepo new_repo.git
cp -r monorepo/.git/refs/replace new_repo.git/refs/replace
cd new_repo.git
git-filter-repo --path aaa --replace-refs delete-no-add
```
Then I have discovered the `--source` and `--target` parameters. I thought they would simplify this to just a single line:

```
git-filter-repo --path aaa --replace-refs delete-no-add --source monorepo --target new_repo.git
```
After making this line work (I had to initialize the target repo).

However this line worked pretty differently. It was much slower 5s->40s, analyzed 20% more commits and the final result had big packfiles while the initial script produced small packs.

What am I doing wrong?

**Comment by newren** [comment]
You're not doing anything wrong, just running into a number of internals that can be collectively suprising.   A simple `git gc --prune=now` after the filter-repo should drop the packfile back down to the expected size.  There's nothing you can do to change the performance.  Some details:

* When you use --source or --target flags, filter-repo can no longer assume the target repo has the same objects as the source repo, and thus cannot pass the --no-data flag to fast-export as an optimization.  This means that all the files are going to need to be serialized and deserialized.  That's expensive.

* filter-repo avoids telling fast-export to filter based on paths, for two reasons: (1) fast-export's path limiting is vastly less flexible so filter-repo would need to filter again anyway, (2) when fast-export filters based on paths there is also history rewriting/limiting going on that can cause problems.  As such, filter-repo just has fast-export send all paths, then filters out what it doesn't want, then passes the rest to fast-import.  So you're paying for serialization of all objects.

* filter-repo doesn't have a way, when it has seen a blob (file contents plus hash, but no name for the file) that it won't be needed later.  It doesn't want to blow memory by attempting to save them all either.  So it passes it along to fast-import as soon as it sees it (unless, e.g. filtering-by-hashes says to remove it).  It'll just be unreferenced and thus pruned out of the packfile at the next gc.  So you're both paying the cost for deserializing all files too, and they appear in the pack.  The size wouldn't be a problem but...

* filter-repo usually runs a `git gc --prune=now` at the end of all runs.  I decided to turn that off when --source or --target were used because you didn't have a mix of new and old commits so we didn't need to worry about mixing history.  I wasn't thinking as much in terms of streamlining all the files, so perhaps it makes sense to put the automatic gc back in there even with those flags.

* And finally, the git clone is fast because you are going from the local filesystem back to itself, so git realizes that it doesn't need to count objects, find out what is reachable, repack things, etc.  In fact, it doesn't even need to copy the objects over -- it can just create some hardlinks to the packfiles and make them all show up instantaneously.  So it saves a massive amount of time in copying the objects over, and because the commands you used got rid of most of them, when you next go to gc (as part of the next filter-repo) it happened to be fast but only because it suddenly became a small repo.

Using --source and --target basically defeats a big pile of optimizations that helped you out for a case like this.  You can fix the big packfiles by doing a `git gc --prune=now` at the end.  I'll consider making that automatic even when --source and --target are passed.

Does this answer your questions?

**Comment by Ark-kun** [comment]
Thank you for thorough explanation (and thank you for a great tool).
This answers my question.

Just couple of thoughts from a user perspective:

I can imagine two kind of filter-like tools - one is expected to run only once with a given target and another kind that can be run with the same source and target repeatedly.

The second tool would be perfect for me (for some time I want to repeatedly update the sub-repo with new commits from the big repo), but I understand that this is probably a rare use case and is probably not in scope for filter-repo.

So, filter-repo seems to be the first kind of tool - it's expected to be only be run once with the given target repo (leaves a maker file and it checks for the bare repo (that started failing for me for some reson... nvm)).
So, if filter-repo wants a clean new target repo, would it make sense for it to create/clone that repo the way it wants? In this case all the assumptions and optimizations are still valid. I expected something like this to happen when `--target` points to a directory that does not exist.  Are there common use cases for `filter-repo` that do not start with something like `git clone --bare`?

**Comment by newren** [comment]
The idea of extracting a sub-piece of a repo and update it as you go is an interesting one.  Closely related to it, is the desire to then take extra development done externally on the sub-repo and bring it back into the main repo via some kind of reversible rewrite.  Once you go down this path you run into a few issues with a tool like filter-repo:

  * Whatever scheme you use to export can't change, otherwise old commits you exported would be exported differently in subsequent runs.  Or you have to keep track of what you've exported across all branches and tags with some kind of mapping so that you know what to export on top of what you've previously exported.  filter-repo doesn't help you with this, and getting it wrong is a recipe for confusion.
  * Many kinds of filtering within filter-repo are non-reversible.  You'd have to limit yourself to a reversible subset.  (See "comments on reversibility" at https://htmlpreview.github.io/?https://github.com/newren/git-filter-repo/blob/docs/html/git-filter-repo.html.)  OR you would have to completely track the mapping between commits in the old and new repositories.  Again, filter-repo doesn't do that, so you'd be doing it on your own and mapping old to new, or else relying on an ability to reverse the rewrite in some way.

While filter-repo has some functionality within this space, it'd be somewhat error prone and frustrating to attempt to use it to fill this need; there's just a bit too much of an impedance mismatch.  However, Christian Schilling at Git Merge 2020 described a tool he was working on that sounds just like the kind of thing you are describing.  He and I had a great discussion there as we discovered we were both writing tools in a similar area, but we had a few different underlying assumptions and usecases; and these differences led to significantly design requirements in the tools.  His tool might be of a lot of interest to you.  See https://git-merge.com/#christian-schilling, https://github.com/initcrash, and https://github.com/esrlabs/josh.

**Comment by newren** [comment]
As mentioned a few comments above, I just pushed a change to make it so that automatic repacking isn't disabled by --source or --target; see commit cdb7b77f07 (filter-repo: repack with --source or --target, 2020-05-28).  As such, I'll go ahead and close this out.
